source/event_stream_rpc_server.c

```diff
@@ -0,0 +1,1107 @@
+    aws_mem_release(message_args->allocator, message_args);
+}
+
+static int s_send_protocol_message(
+    struct aws_event_stream_rpc_server_connection *connection,
+    struct aws_event_stream_rpc_server_continuation_token *continuation,
+    const struct aws_event_stream_rpc_message_args *message_args,
+    int32_t stream_id,
+    aws_event_stream_rpc_server_message_flush_fn *flush_fn,
+    void *user_data) {
+
+    size_t connect_handshake_completion = aws_atomic_load_int(&connection->handshake_complete);
+    AWS_LOGF_TRACE(
+        AWS_LS_EVENT_STREAM_RPC_SERVER,
+        "id=%p: connect handshake state %zu",
+        (void *)connection,
+        connect_handshake_completion);
+    /* handshake step 1 is a connect message being received. Handshake 2 is the connect ack being sent.
+     * no messages other than connect and connect ack are allowed until this count reaches 2. */
+    if (connect_handshake_completion != 2 &&
+        message_args->message_type < AWS_EVENT_STREAM_RPC_MESSAGE_TYPE_CONNECT_ACK) {
+        AWS_LOGF_TRACE(
+            AWS_LS_EVENT_STREAM_RPC_SERVER,
+            "id=%p: invalid state, a message was received prior to connect handshake completion",
+            (void *)connection);
+        return aws_raise_error(AWS_ERROR_EVENT_STREAM_RPC_PROTOCOL_ERROR);
+    }
+
+    struct event_stream_connection_send_message_args *args =
+        aws_mem_calloc(connection->allocator, 1, sizeof(struct event_stream_connection_send_message_args));
+
+    if (!message_args) {
+        AWS_LOGF_ERROR(
+            AWS_LS_EVENT_STREAM_RPC_SERVER,
+            "id=%p: allocation of callback args failed with error %s",
+            (void *)connection,
+            aws_error_debug_str(aws_last_error()));
+        return AWS_OP_ERR;
+    }
+
+    args->allocator = connection->allocator;
+    args->user_data = user_data;
+    args->message_type = message_args->message_type;
+    args->connection = connection;
+    args->flush_fn = flush_fn;
+
+    if (continuation) {
+        args->continuation = continuation;
+        aws_event_stream_rpc_server_continuation_acquire(continuation);
+
+        if (message_args->message_flags & AWS_EVENT_STREAM_RPC_MESSAGE_FLAG_TERMINATE_STREAM) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_EVENT_STREAM_RPC_SERVER,
+                "id=%p: continuation with terminate stream flag was specified closing",
+                (void *)continuation);
+            args->end_stream = true;
+        }
+    }
+
+    if (message_args->message_type == AWS_EVENT_STREAM_RPC_MESSAGE_TYPE_CONNECT_ACK &&
+        !(message_args->message_flags & AWS_EVENT_STREAM_RPC_MESSAGE_FLAG_CONNECTION_ACCEPTED)) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_EVENT_STREAM_RPC_SERVER,
+            "id=%p: connection ack was rejected closing connection",
+            (void *)connection);
+        args->terminate_connection = true;
+    }
+
+    args->flush_fn = flush_fn;
+
+    size_t headers_count = message_args->headers_count + 3;
```

Comment:
This code is identical between client and server AFAICT. Should we share it?

Reply:
not if it means I have to have functions shared that have interleaving branches/types. Long term the types and logic won't always align, so no, I'd prefer we didn't share them